### PR TITLE
Rewrite function invoker.

### PR DESCRIPTION
**Pull request by ericbottard**
Add support for multi I/O functions.
Marshalling/unmarshalling is now fully extensible and handled by SCF.

Only supports streaming via gRPC. http support comes via 
http://github.com/projectriff/streaming-http-adapter-buildpack

### CODE REVIEW

**Comment by fbiville** [comment]
The new invoker doesn't support the [java-boot-uppercase sample](https://github.com/projectriff-samples/java-boot-uppercase/commit/caa079e3a979e2c354a8b167d50f82505728fef3) (`master`'s tip at the time of writing).

Here is how it fails:
```sh
default/bootupper-processor-7c766b4dc7-nlqfk[function]: 08:52:59.030 [main] ERROR o.s.boot.SpringApplication - Application run failed
default/bootupper-processor-7c766b4dc7-nlqfk[function]: org.springframework.context.ApplicationContextException: Failed to start bean 'functionArchiveDeployer'; nested exception is java.lang.IllegalStateException: Failed to deploy archive file:/workspace/
default/bootupper-processor-7c766b4dc7-nlqfk[function]: 	at org.springframework.context.support.DefaultLifecycleProcessor.doStart(DefaultLifecycleProcessor.java:185)
default/bootupper-processor-7c766b4dc7-nlqfk[function]: 	at org.springframework.context.support.DefaultLifecycleProcessor.access$200(DefaultLifecycleProcessor.java:53)
default/bootupper-processor-7c766b4dc7-nlqfk[function]: 	at org.springframework.context.support.DefaultLifecycleProcessor$LifecycleGroup.start(DefaultLifecycleProcessor.java:360)
default/bootupper-processor-7c766b4dc7-nlqfk[function]: 	at org.springframework.context.support.DefaultLifecycleProcessor.startBeans(DefaultLifecycleProcessor.java:158)
default/bootupper-processor-7c766b4dc7-nlqfk[function]: 	at org.springframework.context.support.DefaultLifecycleProcessor.onRefresh(DefaultLifecycleProcessor.java:122)
default/bootupper-processor-7c766b4dc7-nlqfk[function]: 	at org.springframework.context.support.AbstractApplicationContext.finishRefresh(AbstractApplicationContext.java:893)
default/bootupper-processor-7c766b4dc7-nlqfk[function]: 	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:552)
default/bootupper-processor-7c766b4dc7-nlqfk[function]: 	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:747)
default/bootupper-processor-7c766b4dc7-nlqfk[function]: 	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:397)
default/bootupper-processor-7c766b4dc7-nlqfk[function]: 	at org.springframework.boot.SpringApplication.run(SpringApplication.java:315)
default/bootupper-processor-7c766b4dc7-nlqfk[function]: 	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1226)
default/bootupper-processor-7c766b4dc7-nlqfk[function]: 	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1215)
default/bootupper-processor-7c766b4dc7-nlqfk[function]: 	at io.projectriff.invoker.main.EntryPoint.main(EntryPoint.java:35)
default/bootupper-processor-7c766b4dc7-nlqfk[function]: 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
default/bootupper-processor-7c766b4dc7-nlqfk[function]: 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
default/bootupper-processor-7c766b4dc7-nlqfk[function]: 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
default/bootupper-processor-7c766b4dc7-nlqfk[function]: 	at java.base/java.lang.reflect.Method.invoke(Unknown Source)
default/bootupper-processor-7c766b4dc7-nlqfk[function]: 	at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:48)
default/bootupper-processor-7c766b4dc7-nlqfk[function]: 	at org.springframework.boot.loader.Launcher.launch(Launcher.java:87)
default/bootupper-processor-7c766b4dc7-nlqfk[function]: 	at org.springframework.boot.loader.Launcher.launch(Launcher.java:51)
default/bootupper-processor-7c766b4dc7-nlqfk[function]: 	at org.springframework.boot.loader.JarLauncher.main(JarLauncher.java:52)
default/bootupper-processor-7c766b4dc7-nlqfk[function]: Caused by: java.lang.IllegalStateException: Failed to deploy archive file:/workspace/
default/bootupper-processor-7c766b4dc7-nlqfk[function]: 	at org.springframework.cloud.function.deployer.FunctionArchiveDeployer.deploy(FunctionArchiveDeployer.java:105)
default/bootupper-processor-7c766b4dc7-nlqfk[function]: 	at org.springframework.cloud.function.deployer.FunctionDeployerConfiguration$1.start(FunctionDeployerConfiguration.java:98)
default/bootupper-processor-7c766b4dc7-nlqfk[function]: 	at org.springframework.context.support.DefaultLifecycleProcessor.doStart(DefaultLifecycleProcessor.java:182)
default/bootupper-processor-7c766b4dc7-nlqfk[function]: 	... 20 common frames omitted
default/bootupper-processor-7c766b4dc7-nlqfk[function]: Caused by: java.lang.reflect.InvocationTargetException: null
default/bootupper-processor-7c766b4dc7-nlqfk[function]: 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
default/bootupper-processor-7c766b4dc7-nlqfk[function]: 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
default/bootupper-processor-7c766b4dc7-nlqfk[function]: 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
default/bootupper-processor-7c766b4dc7-nlqfk[function]: 	at java.base/java.lang.reflect.Method.invoke(Unknown Source)
default/bootupper-processor-7c766b4dc7-nlqfk[function]: 	at org.springframework.cloud.function.deployer.FunctionArchiveDeployer.launchFunctionArchive(FunctionArchiveDeployer.java:246)
default/bootupper-processor-7c766b4dc7-nlqfk[function]: 	at org.springframework.cloud.function.deployer.FunctionArchiveDeployer.deploy(FunctionArchiveDeployer.java:78)
default/bootupper-processor-7c766b4dc7-nlqfk[function]: 	... 22 common frames omitted
default/bootupper-processor-7c766b4dc7-nlqfk[function]: Caused by: org.springframework.context.ApplicationContextException: Unable to start web server; nested exception is java.lang.NoClassDefFoundError: org/ietf/jgss/GSSException
default/bootupper-processor-7c766b4dc7-nlqfk[function]: 	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.onRefresh(ServletWebServerApplicationContext.java:155)
default/bootupper-processor-7c766b4dc7-nlqfk[function]: 	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:540)
default/bootupper-processor-7c766b4dc7-nlqfk[function]: 	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:140)
default/bootupper-processor-7c766b4dc7-nlqfk[function]: 	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:775)
default/bootupper-processor-7c766b4dc7-nlqfk[function]: 	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:397)
default/bootupper-processor-7c766b4dc7-nlqfk[function]: 	at org.springframework.boot.SpringApplication.run(SpringApplication.java:316)
default/bootupper-processor-7c766b4dc7-nlqfk[function]: 	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1260)
default/bootupper-processor-7c766b4dc7-nlqfk[function]: 	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1248)
default/bootupper-processor-7c766b4dc7-nlqfk[function]: 	... 28 common frames omitted
default/bootupper-processor-7c766b4dc7-nlqfk[function]: Caused by: java.lang.NoClassDefFoundError: org/ietf/jgss/GSSException
default/bootupper-processor-7c766b4dc7-nlqfk[function]: 	at org.apache.catalina.startup.Tomcat.createDefaultRealm(Tomcat.java:676)
default/bootupper-processor-7c766b4dc7-nlqfk[function]: 	at org.apache.catalina.startup.Tomcat.getEngine(Tomcat.java:519)
default/bootupper-processor-7c766b4dc7-nlqfk[function]: 	at org.apache.catalina.startup.Tomcat.getHost(Tomcat.java:496)
default/bootupper-processor-7c766b4dc7-nlqfk[function]: 	at org.springframework.boot.web.embedded.tomcat.TomcatServletWebServerFactory.getWebServer(TomcatServletWebServerFactory.java:168)
default/bootupper-processor-7c766b4dc7-nlqfk[function]: 	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.createWebServer(ServletWebServerApplicationContext.java:179)
default/bootupper-processor-7c766b4dc7-nlqfk[function]: 	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.onRefresh(ServletWebServerApplicationContext.java:152)
default/bootupper-processor-7c766b4dc7-nlqfk[function]: 	... 35 common frames omitted
default/bootupper-processor-7c766b4dc7-nlqfk[function]: Caused by: java.lang.ClassNotFoundException: org.ietf.jgss.GSSException
default/bootupper-processor-7c766b4dc7-nlqfk[function]: 	at java.base/java.net.URLClassLoader.findClass(Unknown Source)
default/bootupper-processor-7c766b4dc7-nlqfk[function]: 	at java.base/java.lang.ClassLoader.loadClass(Unknown Source)
default/bootupper-processor-7c766b4dc7-nlqfk[function]: 	at org.springframework.boot.loader.LaunchedURLClassLoader.loadClass(LaunchedURLClassLoader.java:92)
default/bootupper-processor-7c766b4dc7-nlqfk[function]: 	at org.springframework.cloud.function.deployer.FunctionArchiveDeployer$1.loadClass(FunctionArchiveDeployer.java:162)
default/bootupper-processor-7c766b4dc7-nlqfk[function]: 	... 41 common frames omitted
```

My initial reaction was to [remove the unneeded dependencies](https://github.com/projectriff-samples/java-boot-uppercase/pull/4) in the sample. 
However, as @trisberg pointed out, these are valid dependencies in general and that should not trigger the above error.

You can reproduce this with the following script:
```sh
#!/usr/bin/env bash
set -Eeuo pipefail

function delete_processor() {
  riff streaming processor delete "${1}" &> /dev/null || true
}

function delete_function() {
  riff function delete "${1}" &> /dev/null || true
}

function delete_stream() {
   riff streaming stream delete "${1}" &> /dev/null || true
}

function create_stream() {
  riff streaming stream create "${1}" --content-type "${2}" --provider kafka-provider
}

function create_function() {
  riff function create "${1}" --git-repo "${2}" --tail
}

function create_processor() {
  name="${1}"
  function="${2}"
  shift
  shift
  riff streaming processor create "${name}" --function-ref "${function}" "${@}"
}

function showProtip() {
  echo '💡 Feel free to run "watch kubectl get function,processor,stream,pod" in a separate terminal'
  sleep 5
}

function tearDown() {
  echo '💥 Deleting existing demo processors if any'
  delete_processor bootupper

  echo '💥 Deleting existing demo functions if any'
  delete_function bootupper

  echo '💥 Deleting existing demo streams if any'
  delete_stream letters
  delete_stream upper
}

function setUp() {
  echo '🎉 Creating streams'
  create_stream letters 'text/plain'
  create_stream upper 'application/json'

  echo '🎉 Creating functions'
  create_function bootupper 'https://github.com/projectriff-samples/java-boot-uppercase.git'

  echo '🎉 Creating processors'
  create_processor bootupper bootupper --input repeated --output upper

  echo '⌨️  Now open 2 terminals, move to your liiklus-client clone and run these (1 per terminal):'
  echo java -jar target/liiklus-client-1.0.0-SNAPSHOT.jar --producer "$(kubectl get svc liiklus -ojsonpath='{.status.loadBalancer.ingress[0].ip}'):6565" letters
  echo java -jar target/liiklus-client-1.0.0-SNAPSHOT.jar --consumer "$(kubectl get svc liiklus -ojsonpath='{.status.loadBalancer.ingress[0].ip}'):6565" upper
}

showProtip
tearDown
setUp
```

**Comment by fbiville** [comment]
**TL;DR** 
I assume the latest Spring Cloud Function Deployer is not meant to be compatible with the `2.0.0.RC2` previous version. Do you confirm @ericbottard @olegz?


Reproduced in `testMulti`:

```sh
cd $(mktemp -d) && \
    git clone https://github.com/fbiville/java-function-invoker.git -b uppercase-deps && \
    cd java-function-invoker && \
    mvn -Dit.test='IntegrationTest#testMulti' verify
```

```java
11:35:04.432 [main] ERROR o.s.boot.SpringApplication - Application run failed
org.springframework.context.ApplicationContextException: Failed to start bean 'functionArchiveDeployer'; nested exception is java.lang.IllegalStateException: Failed to deploy archive file:/Users/fbiville/workspace/java-function-invoker/src/test/functions/repeater-as-bean-1.0.0-boot.jar
	at org.springframework.context.support.DefaultLifecycleProcessor.doStart(DefaultLifecycleProcessor.java:185)
	at org.springframework.context.support.DefaultLifecycleProcessor.access$200(DefaultLifecycleProcessor.java:53)
	at org.springframework.context.support.DefaultLifecycleProcessor$LifecycleGroup.start(DefaultLifecycleProcessor.java:360)
	at org.springframework.context.support.DefaultLifecycleProcessor.startBeans(DefaultLifecycleProcessor.java:158)
	at org.springframework.context.support.DefaultLifecycleProcessor.onRefresh(DefaultLifecycleProcessor.java:122)
	at org.springframework.context.support.AbstractApplicationContext.finishRefresh(AbstractApplicationContext.java:893)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:552)
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:747)
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:397)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:315)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1226)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1215)
	at io.projectriff.invoker.main.EntryPoint.main(EntryPoint.java:35)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:48)
	at org.springframework.boot.loader.Launcher.launch(Launcher.java:87)
	at org.springframework.boot.loader.Launcher.launch(Launcher.java:51)
	at org.springframework.boot.loader.JarLauncher.main(JarLauncher.java:52)
Caused by: java.lang.IllegalStateException: Failed to deploy archive file:/Users/fbiville/workspace/java-function-invoker/src/test/functions/repeater-as-bean-1.0.0-boot.jar
	at org.springframework.cloud.function.deployer.FunctionArchiveDeployer.deploy(FunctionArchiveDeployer.java:105)
	at org.springframework.cloud.function.deployer.FunctionDeployerConfiguration$1.start(FunctionDeployerConfiguration.java:105)
	at org.springframework.context.support.DefaultLifecycleProcessor.doStart(DefaultLifecycleProcessor.java:182)
	... 20 common frames omitted
Caused by: java.lang.reflect.InvocationTargetException: null
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.springframework.cloud.function.deployer.FunctionArchiveDeployer.launchFunctionArchive(FunctionArchiveDeployer.java:249)
	at org.springframework.cloud.function.deployer.FunctionArchiveDeployer.deploy(FunctionArchiveDeployer.java:78)
	... 22 common frames omitted
Caused by: org.springframework.context.ApplicationContextException: Unable to start web server; nested exception is java.lang.NoClassDefFoundError: org/ietf/jgss/GSSException
	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.onRefresh(ServletWebServerApplicationContext.java:155)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:540)
	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:140)
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:775)
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:397)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:316)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1260)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1248)
	... 28 common frames omitted
Caused by: java.lang.NoClassDefFoundError: org/ietf/jgss/GSSException
	at org.apache.catalina.startup.Tomcat.createDefaultRealm(Tomcat.java:676)
	at org.apache.catalina.startup.Tomcat.getEngine(Tomcat.java:519)
	at org.apache.catalina.startup.Tomcat.getHost(Tomcat.java:496)
	at org.springframework.boot.web.embedded.tomcat.TomcatServletWebServerFactory.getWebServer(TomcatServletWebServerFactory.java:168)
	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.createWebServer(ServletWebServerApplicationContext.java:179)
	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.onRefresh(ServletWebServerApplicationContext.java:152)
	... 35 common frames omitted
Caused by: java.lang.ClassNotFoundException: org.ietf.jgss.GSSException
	at java.base/java.net.URLClassLoader.findClass(URLClassLoader.java:471)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:588)
	at org.springframework.boot.loader.LaunchedURLClassLoader.loadClass(LaunchedURLClassLoader.java:92)
	at org.springframework.cloud.function.deployer.FunctionArchiveDeployer$1.loadClass(FunctionArchiveDeployer.java:162)
	... 41 common frames omitted
```

**Comment by fbiville** [comment]
Interestingly, the same problem still happens even if the function pulls the same Spring Cloud Function dependency versions as the invoker (see the diff applied against the sample to achieve that):
```
diff --git a/pom.xml b/pom.xml
index bf3bca9..61cc5ca 100644
--- a/pom.xml
+++ b/pom.xml
@@ -22,7 +22,7 @@
 		<project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
 		<project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
 		<java.version>1.8</java.version>
-		<spring-cloud.version>Greenwich.M3</spring-cloud.version>
+		<spring-cloud-function.version>3.0.0.BUILD-SNAPSHOT</spring-cloud-function.version>
 	</properties>
 
 	<dependencies>
@@ -42,17 +42,17 @@
 		</dependency>
 	</dependencies>
 
-	<dependencyManagement>
-		<dependencies>
-			<dependency>
-				<groupId>org.springframework.cloud</groupId>
-				<artifactId>spring-cloud-dependencies</artifactId>
-				<version>${spring-cloud.version}</version>
-				<type>pom</type>
-				<scope>import</scope>
-			</dependency>
-		</dependencies>
-	</dependencyManagement>
+    <dependencyManagement>
+        <dependencies>
+            <dependency>
+                <groupId>org.springframework.cloud</groupId>
+                <artifactId>spring-cloud-function-dependencies</artifactId>
+                <version>${spring-cloud-function.version}</version>
+                <type>pom</type>
+                <scope>import</scope>
+            </dependency>
+        </dependencies>
+    </dependencyManagement>
 
 	<build>
 		<plugins>
@@ -64,6 +64,17 @@
 	</build>
 
 	<repositories>
+		<repository>
+			<id>spring-snapshots</id>
+			<name>Spring Snapshots</name>
+			<url>https://repo.spring.io/libs-snapshot-local</url>
+			<snapshots>
+				<enabled>true</enabled>
+			</snapshots>
+			<releases>
+				<enabled>false</enabled>
+			</releases>
+		</repository>
 		<repository>
 			<id>spring-milestones</id>
 			<name>Spring Milestones</name>

```

**Comment by olegz** [comment]
This is indeed a bit strange especially given the when Eric and I worked on it we used the `repeater-as-bean-1.0.0-boot.jar` and most of existing riff apps for validation.
That said, this should be easy to debug (why it can not be found) once we (I) understand where is `org.ietf.jgss.GSSException` and who is looking for it.

Ping me when you get a min so we can chat.

With regard to - `not meant to be compatible with the 2.0.0.RC2 previous version` - that depends on how you look at it. Internally deployer does the same thing as it always done, so behaviourally nothing has changed, however interaction with it is now simplified where JavaInvoker is really just a boot app. You inject `FunctionCatalog` into whatever you want to inject it and run with it, so yes there are no more AbstractBotstrap and other `special` classes.

**Comment by olegz** [comment]
Also, one thing to mention that is quite important. The spring-cloud-function 2.x is based on boot 2.1 while spring-cloud-function 3.x (which is where new deployer is) is based on boot 2.2.
As you can see from our [latest 3.0.0.M2 announcement](https://spring.io/blog/2019/08/15/announcing-spring-cloud-function-3-0-0-m2) we had to do 3.0.0.M1 just to establish compatibility with boot 2.2.

File: pom.xml [hunk 1]
```diff
@@ -1,263 +1,321 @@
...
-		<project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
-		<java.version>11</java.version>
-		<spring-cloud-function.version>2.0.1.RELEASE</spring-cloud-function.version>
-		<skip.invoke>true</skip.invoke>
-	</properties>
+    <properties>
+        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+        <project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
+        <java.version>11</java.version>
+        <grpc.version>1.15.1</grpc.version>
```
Comment (fbiville) [inline pom.xml]: Please upgrade to [`1.23.0`](https://github.com/grpc/grpc-java/releases/tag/v1.23.0) as this release contains fixes for [HTTP2 vulnerabilities](https://www.kb.cert.org/vuls/id/605641/).

Comment (fbiville) [reply pom.xml]: Done with https://github.com/projectriff/java-function-invoker/pull/133/commits/9c4fa0b35e6903577609912a3d85a1cbbe223cb6.

File: src/main/java/io/projectriff/invoker/client/FunctionClient.java [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+import java.util.function.Function;
+
+/**
+ * FunctionClient is a client-side helper class to invoke riff streaming function over gRPC.
+ *
+ * <p>It performs client-side serialization/deserializaton and mux/demux of parameters and return values.</p>
+ * <p>By default, only a converter for {@code application/json} is set up, but users can override this via {@link #setMessageConverters(AbstractMessageConverter...)}.</p>
+ *
+ * @param <I> The input type of the function, typycally {@code Flux<T>} or {@code TupleN<Flux<T>, Flux<U>, ...>}
+ * @param <>  The output type of the function, typycally {@code Flux<R>} or {@code TupleM<Flux<R>, Flux<S>, ...>}
```
Comment (fbiville) [inline src/main/java/io/projectriff/invoker/client/FunctionClient.java]:
```suggestion
 * @param <O>  The output type of the function, typycally {@code Flux<R>} or {@code TupleM<Flux<R>, Flux<S>, ...>}
```

File: src/main/java/io/projectriff/invoker/main/EntryPoint.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+import org.springframework.context.ApplicationContext;
+import org.springframework.context.SmartLifecycle;
+import org.springframework.context.annotation.Bean;
+
+import java.io.IOException;
+
+/**
+ * This is the main entry point for the java function invoker.
+ * This sets up an application context with the whole Spring Cloud Function infrastructure (thanks to auto-configuration)
+ * set up, pointing to the user function (via correctly set {@link FunctionProperties} ConfigurationProperties.
```
Comment (fbiville) [inline src/main/java/io/projectriff/invoker/main/EntryPoint.java]:
```suggestion
 * setup, pointing to the user function (via correctly set {@link FunctionProperties} ConfigurationProperties.
```

**Comment by fbiville** [comment]
Merged in 119b922ea957ec998c48a564fb201550dcc2f7bf.

**Pull request closed without merging**